blockchain/validator_set.go

```diff
@@ -0,0 +1,273 @@
+package blockchain
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"math"
+	"sort"
+	"strconv"
+
+	acm "github.com/hyperledger/burrow/account"
+	rpc_core "github.com/tendermint/tendermint/rpc/core"
+	ctypes "github.com/tendermint/tendermint/rpc/core/types"
+	tm_types "github.com/tendermint/tendermint/types"
+)
+
+type ValidatorSet interface {
+	TotalPower() int
+	SetMaximumPower(maximumPower int)
+	MaximumPower() int
+	AdjustPower(height int64) error
+	Validators() []acm.Validator
+	SetLeavers() []acm.Validator
+	JoinToTheSet(validator acm.Validator) error
+	LeaveFromTheSet(validator acm.Validator) error
+	IsValidatorInSet(address acm.Address) bool
+	JSONBytes() ([]byte, error)
+}
+
+type validatorListProxy interface {
+	Validators(height int64) (*ctypes.ResultValidators, error)
+}
+
+type _validatorListProxy struct{}
+
+func (vlp _validatorListProxy) Validators(height int64) (*ctypes.ResultValidators, error) {
+	result, err := rpc_core.Validators(&height)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+type validatorSet struct {
+	maximumPower int
+	validators   []acm.Validator
+	setLeavers   []acm.Validator
+	proxy        validatorListProxy
+}
+
+func newValidatorSet(maximumPower int, validators []acm.Validator) *validatorSet {
+	set := &validatorSet{
+		validators: validators,
+		proxy:      _validatorListProxy{},
+	}
+	set.SetMaximumPower(maximumPower)
+	return set
+}
+
+func (vs *validatorSet) TotalPower() int {
+	return len(vs.validators)
+}
+
+func (vs *validatorSet) SetMaximumPower(maximumPower int) {
+	if maximumPower > 90 {
+		maximumPower = 90
+	}
+
+	if maximumPower < 4 {
+		maximumPower = 4
+	}
+
+	vs.maximumPower = maximumPower
+}
+
+func (vs *validatorSet) MaximumPower() int {
+	return vs.maximumPower
+}
+
+func (vs *validatorSet) AdjustPower(height int64) error {
```

Comment:
This function could do with some explanatory comments

Reply:
You are right. I need to put more comments here. 
This algorithm is a little complicated and I tried to write it as simple as possible. If you look at the code, the main loop it's almost around 40 lines. The main idea behind this algorithm is comparing current validator set with previous sets. Any validator who is the current set but not in the previous set(s) has joined recently. I try to sort arrays before comparing them, so it helps to compare arrays easier and faster (the performance is `O(n log(n))`)... 
However, this piece of the code is very important in order to keep blockchain alive. Any mistakes here will cause blockchain goes dead.
About 1/3, The chance of changing 1/3 of the nodes in each round is almost zero. To make a blockchain safe and fast enough we figured out that a number between 60 and 80 validators is the best for validator set. (maximum power should be something between 60~80)
Take 60 as an example, It's almost impossible that 20 nodes nominating themselves to enter to the validator set in one round. Almost impossible! Impossible is possible, so we always make sure to not change more than 2/3 of nodes in each round. 
How about any possible attack? It's possible. Let me explain it a little more here.

Consider the VRF method is working fine and the byzantine party have no control over it. (we need to consult with some cryptographers to make sure about its safety. I believe we can make it safe with including some cure parameters into the proof. I am not worried about this part.)
If the VRF itself is safe, the only way to attack the blokchain is generating so many random pair keys and try to find suitable values for sending sortition transaction.
I try to prevent this attack by two changes:
1- The seed for the VRF in each round is almost random and unpredictable; It's previous block hash.
2- There is a gap between first bonding transaction and first sortition transaction.

I figured out that in a live network it might not possible for a node to send a sortition transaction just for entering into the next validation set. It might not be seen by every node at coming round. This transaction should be valid for more than one round, But it can not be valid forever. My suggestion is that the sortition transaction can be valid for just 10 rounds (I am very keen for changing it).
This is the only way an attacker can attack the blockchain. They can start generating valid sortition transaction by using lots of random generated key pairs. Since this sortiton transaction is valid for 10 rounds they can attack the blockchain in between.
The protocol predicts this attack, and we prevented it very simply.
Before a node starts sending a sortition transaction, it should bond some amount. Sortition transaction can only send 10 rounds after bonding transaction. After bonding transaction the node can not send any sortiton for first 10 rounds.

Reply:
@seanyoung and I will be taking some time to study you design paper and look a this. 

I think you are right to identify what is effectively a stake grinding attack. This is how proof of work protocols can degenerate into proof of work. The usual mitigation for this is to slow down the rate at which stake can move around, which is effectively what I think you are saying if that you have to stake to a key pair before you can know the block hash which would allow you to throw computation at finding a valid sortition address/key.

I like this approach you are taking - but I do think it needs a careful pencil and paper analysis by a computer scientist/cryptographer.